### DISCUSSION TITLE
Noob question : avoid reuploading data between two kernel calls

### DISCUSSION THREAD

**Discussion by mikamyara** [pyopencl/pyopencl#692] [Q&A] [answered]
Dear all,
  I have two kernels, which prototypes can be, for example :
```
  void k1(__global float* input, __global float* output);
  void k2(__global float* input, __global float* output);
```
At this time, to call them, I do something like : (queue is a queue, ctx a context, prg the compiled kernels)
```
array1_np = np.zeros((100,100),dtype=np.float32)  # + put some data inside array1_np
array1_g= cl.Buffer(ctx, mf.READ_ONLY | mf.COPY_HOST_PTR, hostbuf=image_np)
array2_g = cl.Buffer(ctx, mf.READ_WRITE, 1000*1000*4)
array3_g = cl.Buffer(ctx, mf.READ_WRITE, 10*10*4)

k1 = prg.k1
k1(queue,(100,100),None,array1_g,array2_g)
k2 = prg.k2
k2(queue,(1000,1000),None,array2_g,array3_g)
```

and then some queue code to download array2_g and array3_g.

I wanted to be sure about it : as there is no "COPY_HOST_PTR" when queuing array2_g and array3_g,  can I be sure that array2_g will not be uploaded again when k2 will be called ?
Also : I would like to know when exactly the code of k1 and k2 is run : is it at the moment at which k1 and k2 are queued, or is it when the outputs are downloaded ?
Hope my questions make sense ...

Best regards,
   Mikhaël

**Comment by inducer** [comment, marked as answer]
> can I be sure that array2_g will not be uploaded again when k2 will be called ?

Yes. Even for `array1_g`, the copy happens on buffer creation and not thereafter.

> I would like to know when exactly the code of k1 and k2 is run : is it at the moment at which k1 and k2 are queued, or is it when the outputs are downloaded ?

- OpenCL does not define when the code runs as written.
- If the transfers were not synchronous (as yours are) but instead enqueued (`cl.enqueue_copy(..., is_blocking=False)`), then the queue would guarantee execution in sequence.
- If you would like to force the code to run, you could call `queue.finish()`.

**Reply by mikamyara** [reply]
Very clear anwser. Thanks a lot !
